lib/xform.js

```diff
@@ -41,18 +55,159 @@ var XFormState = State.extend(function (self, name, opts) {
+                self.xforms_service_url + '/responses/' + xform_id + '/0')
+            .then(function(resp) {
+                return resp.data.question;
+            }, function(err) {
+            return self.im.log.error([
+                "HTTP Error in connecting to the xforms service",
+                err.message])
+                .then(function() {
+                    self.metadata.display_text = self.xforms_service_error_message;
+                    self.metadata.xforms_state = null;
+                    return false;
+                });
+        });
+    };
+
+    self.handlers.initial = function(content) {
+        if(typeof(self.xform) === 'function') {
+            self.xform = self.xform();
+        }
+        return Q(self.xform)
+            .then(self.create_xform)
+            .then(self.get_first_xform_question)
+            .then(function(question) {
+                if(!question) {
+                    return;
+                }
+                self.metadata.display_text = question;
+                self.metadata.xformstate = 'next_question';
+                return;
+            });
+    };
+
+    self.answer_question = function(answer) {
+        http = new JsonApi(self.im);
+        return http.post(
+                self.xforms_service_url + '/responses/' +
+                self.metadata.xform_id,
+                {
+                    data: {
+                        'answer': answer,
+                    }
+                }
+            ).then(function(resp) {
+                return resp.data.question;
+            }, function(err) {
+            return self.im.log.error([
+                "HTTP Error in connecting to the xforms service",
+                err.message])
+                .then(function() {
+                    self.metadata.display_text = self.xforms_service_error_message;
+                    self.metadata.xforms_state = null;
+                    return false;
+                });
+        });
+    };
+
+    self.set_next_question = function(question) {
+        if(question === false) {
+            return;
+        }
+        if(!question) {
+            return self.get_responses()
+                .then(function(responses) {
+                    return self.set_next_state(self.next, responses);
+                });
+        }
+        self.metadata.display_text = question;
+        self.metadata.xformstate = 'next_question';
+        return;
+    };
+
+    self.get_responses = function() {
+        var http = new HttpApi(self.im);
+        return http.get(
+            self.xforms_service_url + '/answers/' + self.metadata.xform_id)
+            .then(function(resp) {
+                return resp.data;
+            });
+    };
+
+    self.handlers.next_question = function(content) {
+        return Q(content)
+            .then(self.answer_question)
+            .then(self.set_next_question);
+    };
+
+    self.display = function() {
+        return self.metadata.display_text || '';
```

Comment:
Probably we should never display nothing. Any thoughts on what to display instead? What are the cases where `display_text` can not exist?